### PR TITLE
(PDB-2242) Move from moustache to comidi

### PR DESCRIPTION
**Pull request by senior**
This series of commits moves PuppetDB from moustache to comidi and refactors routes and handlers. All query routes are centralized in a single namespace, parameterized by version.

More refactoring work needs to be done. Specifically there's some rough edges around adding the parent check to routes and how URL decoding route segments is done.

Upgrading comidi to the latest (0.3.1) also required updating some other depdencies which is included in this PR as well.

### CODE REVIEW

File: test/puppetlabs/puppetdb/metrics/core_test.clj [hunk 1]
```diff
@@ -72,6 +72,7 @@
                     :let [response (app
                                     (get-request
                                      (str version uri)))]]
+              (println                        (str version uri))
```
Comment (kbarber) [inline test/puppetlabs/puppetdb/metrics/core_test.clj]: debugging

Comment (kbarber) [inline test/puppetlabs/puppetdb/metrics/core_test.clj]: sorry this is wip isn't it, carry on :-).

File: src/puppetlabs/puppetdb/http/command.clj [hunk 2]
```diff
@@ -6,12 +6,12 @@
             [puppetlabs.puppetdb.command :as command]
             [puppetlabs.puppetdb.http :as http]
             [puppetlabs.puppetdb.cheshire :as json]
-            [net.cgrand.moustache :as moustache]
             [puppetlabs.puppetdb.config :as conf]
             [puppetlabs.puppetdb.middleware :as mid]
             [compojure.core :as compojure]
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/command.clj]: This isn't used anymore

File: src/puppetlabs/puppetdb/meta.clj [hunk 3]
```diff
@@ -1,16 +1,18 @@
 (ns puppetlabs.puppetdb.meta
   (:require [clojure.tools.logging :as log]
-            [net.cgrand.moustache :as moustache]
             [clj-time.core :refer [now]]
-            [puppetlabs.puppetdb.middleware
-             :refer [verify-accepts-json wrap-with-globals validate-no-query-params]]
+            [puppetlabs.puppetdb.middleware :as mid]
             [puppetlabs.trapperkeeper.core :refer [defservice]]
             [compojure.core :as compojure]
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/meta.clj]: This isn't used anymore

File: project.clj [hunk 4]
```diff
@@ -65,15 +65,14 @@
                  ;; bridge to allow some spring/activemq stuff to log over slf4j
                  [org.slf4j/jcl-over-slf4j "1.7.10"]
                  ;; WebAPI support libraries.
-                 [net.cgrand/moustache "1.1.0" :exclusions [ring/ring-core org.clojure/clojure]]
-                 [compojure "1.3.3"]
+                 [compojure "1.4.0"]
```
Comment (ajroetker) [inline project.clj]: Can we remove this dependency now?

File: src/puppetlabs/puppetdb/http/handlers.clj [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+   
+   (cmdi/ANY ["/" :hash "/metrics"] []
+             (-> (report-data-responder version "report_metrics")
+                 (comp decode-route-params)
+                 (parent-check version :report :hash)
+                 validate-no-query-params))
+   
+   (cmdi/ANY ["/" :hash "/logs"] []
+             (-> (report-data-responder version "report_logs")
+                 (comp decode-route-params)
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/handlers.clj]: seems like decode-route-params should just be part of `wrap-with-puppetdb-middleware`, i.e. a global middleware

File: src/puppetlabs/puppetdb/http/handlers.clj [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+  ([param-spec :- params-schema
+    routes :- bidi-schema/RoutePair]
+   (cmdi/wrap-routes routes #(http-q/extract-query % param-spec))))
+
+(pls/defn-validated append-handler :- bidi-schema/RoutePair
+  [route :- bidi-schema/RoutePair
+   handler-to-append :- handler-schema]
+  (cmdi/wrap-routes route #(comp % handler-to-append)))
+
+(pls/defn-validated create-handler :- handler-schema
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/handlers.clj]: could we call this create-query-handler, or even just query-handler?

File: src/puppetlabs/puppetdb/middleware.clj [hunk 7]
```diff
@@ -342,18 +344,37 @@
...
+  ([route handler-fn]
+      (assert route "Cannot create a Ring handler with a nil Route(s) parameter")
+      (fn [{:keys [uri path-info] :as req}]
+        (let [path (or path-info uri)
+              {:keys [handler route-params] :as match-context}
+              (bidi/match-route* route path req)]
+          (when handler
+            (bring/request
+             (handler-fn handler)
+             (update-in req [:route-params] merge route-params)
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/middleware.clj]: `(update req :route-params...`

**Comment by ajroetker** [comment]
Lgtm for the most part but FYI this causes PE to not be able to compile, otherwise :+1:

File: project.clj (hunk 4 above)
Comment (senior) [inline project.clj]: Not yet. I forgot that we still use some stuff from compojure. pdb-routing.clj is a good example. I don't see anything we're using compojure for that wouldn't be pretty easy to move over to comidi. Once we've done that, we'll have no code that references compojure and I'll feel comfortable removing it.

File: src/puppetlabs/puppetdb/http/handlers.clj (hunk 5 above)
Comment (senior) [inline src/puppetlabs/puppetdb/http/handlers.clj]: I wasn't confident that the middleware would be applied at the right time, I can try it out though.

**Comment by senior** [comment]
@ajroetker I have a PR up for the PE side, it'll need to be updated once I push the change of create-handler -> create-query handler, but works as-is now. I'll investigate the decode params thing in the morning.

**Comment by senior** [comment]
@ajroetker Addressed your comments. Renamed `create-handler` to `create-query-handler` and made the related updates. URL decoding route-params in a middleware didn't work but I did modify make-pdb-handler to do the right thing and decode it there. Looking at that function now, seems like the right place for it anyway.

I made a separate commit with the changes so it should be easy to review on it's own.

**Comment by kbarber** [comment]
@senior let me do some poking at this now.

File: src/puppetlabs/puppetdb/http/v4.clj [hunk 8]
```diff
@@ -1,75 +1,24 @@
...
-
-   ["event-counts" &]
-   {:any (ec/event-counts-app version)}
-
-   ["aggregate-event-counts" &]
-   {:any (aec/aggregate-event-counts-app version)}
-
-   ["reports" &]
-   {:any (reports/reports-app version)}))
+  (apply cmdi/routes
```
Comment (kbarber) [inline src/puppetlabs/puppetdb/http/v4.clj]: Why not collapse this file completely, into server.clj? Then we can see v5 routes next to v4 routes next to each other :-).

File: src/puppetlabs/puppetdb/meta.clj (hunk 3 above)
Comment (kbarber) [inline src/puppetlabs/puppetdb/meta.clj]: Indeed.

File: test/puppetlabs/puppetdb/acceptance/cli.clj [hunk 9]
```diff
@@ -15,7 +15,7 @@
             [puppetlabs.puppetdb.testutils.tar :refer [tar->map]]))
 
 (deftest test-anonymized-export-roundtrip
-  (doseq [profile (keys anon/anon-profiles)]
+  (doseq [profile  (take 1 (keys anon/anon-profiles))]
```
Comment (kbarber) [inline test/puppetlabs/puppetdb/acceptance/cli.clj]: You'll be incurring whitespace tax here, lol

Comment (senior) [inline test/puppetlabs/puppetdb/acceptance/cli.clj]: No idea why this change was made, I'll roll it back

File: src/puppetlabs/puppetdb/http/v4.clj (hunk 8 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/v4.clj]: :+1:  to that, I never liked the namespace per api version convention

**Comment by kbarber** [comment]
@senior 3 things:

* That cli.clj thingy
* Consider collapsing v4 -> server.clj
* The left over reference to compojure that AJ mentioned

Otherwise I'm +1, the new layout feels good to me, compressing this stuff into handlers.clj means the problem is together, and we can start to apply code reduction easier now. Well done.

File: project.clj (hunk 4 above)
Comment (ajroetker) [inline project.clj]: Should we ticket that work then for another patch?

Comment (kbarber) [inline project.clj]: I think another patch/ticket is a good idea, we definitely want to close the book on this soon though.

**Comment by senior** [comment]
@kbarber @ajroetker v4 is now collapsed into server.clj, that stray compojure reference has been removed, same with the inadvertant cli test change

**Pull request merged**